File: Source/Engine/UI/Experimental/Primitive/Private/InputEventHook.cpp
//                                                        UI System
//                                                    writen by Nori_SC
//                                                https://github.com/NoriteSC
#include "../InputEventHook.h"
#include "../../../../Debug/DebugLog.h"
#include <Engine/Scripting/Internal/InternalCalls.h>
UIInputEventHook::UIInputEventHook(const SpawnParams& params) : UIComponent(params){}

UIEventResponse UIInputEventHook::OnPointerInput(const UIPointerEvent& InEvent)
{
    if (PointerInput.IsBinded())
    {
        PointerInput(InEvent);
    }
    
    return UIEventResponse::None;
}

UIEventResponse UIInputEventHook::OnActionInput(const UIActionEvent& InEvent)
{
    if (ActionInput.IsBinded())
    {
        ActionInput(InEvent);
    }
    return UIEventResponse::None;
}

 void UIInputEventHook::CSHACK_SetActionInputUIEventResponse(const UIEventResponse& InEventResponse)
{
    CSHACK_ActionInputUIEventResponse = InEventResponse;
}

 void UIInputEventHook::CSHACK_SetPointerInputUIEventResponse(const UIEventResponse& InEventResponse)
{
    CSHACK_PointerInputUIEventResponse = InEventResponse;
}

void UIInputEventHook::CSHACK_BindActionInput()
{
    if (ActionInput.IsBinded())
        ActionInput.Unbind();
    ActionInput.Bind([this](const UIActionEvent& InEvent) {return CSHACK_InvokeActionInput(InEvent); });
}

void UIInputEventHook::CSHACK_BindPointerInput()
{
    if (PointerInput.IsBinded())
        PointerInput.Unbind();
    PointerInput.Bind([this](const UIPointerEvent& InEvent) {return CSHACK_InvokePointerInput(InEvent); });
}

void UIInputEventHook::CSHACK_UnBindActionInput()
{
    if (ActionInput.IsBinded())
        ActionInput.Unbind();
}

void UIInputEventHook::CSHACK_UnBindPointerInput()
{
    if (PointerInput.IsBinded())
        PointerInput.Unbind();
}

bool UIInputEventHook::CSHACK_IsBindedActionInput()
{
    return ActionInput.IsBinded();
}

bool UIInputEventHook::CSHACK_IsBindedPointerInput()
{
    return PointerInput.IsBinded();
}

UIEventResponse UIInputEventHook::CSHACK_InvokePointerInput(const UIPointerEvent& InEvent)
{
    CSHACK_PointerInput(InEvent);
    return CSHACK_PointerInputUIEventResponse;
}

UIEventResponse UIInputEventHook::CSHACK_InvokeActionInput(const UIActionEvent& InEvent)
{
    CSHACK_ActionInput(InEvent);
    return CSHACK_ActionInputUIEventResponse;
}

File: Source/Engine/UI/Experimental/Primitive/InputEventHook.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace FlaxEngine.Experimental.UI
{
    public partial class UIInputEventHook
    {
        public Func<UIActionEvent, UIEventResponse> ActionInput
        {
            get => m_ActionInput;
            set
            {
                m_ActionInput = value;
                HandleActionInputCPPBindings();
            }
        }

        public Func<UIPointerEvent, UIEventResponse> PointerInput
        {
            get => m_PointerInput;
            set
            {
                m_PointerInput = value;
                HandlePointerInputCPPBindings();
            }
        }

        #region CSHACK

        private Func<UIPointerEvent, UIEventResponse> m_PointerInput;
        private Action<UIPointerEvent> m_PointerInputAction;
        private Func<UIActionEvent, UIEventResponse> m_ActionInput;
        private Action<UIActionEvent> m_ActionInputAction;
        private void HandlePointerInputCPPBindings()
        {
            if (m_PointerInput == null)
            {
                CSHACK_PointerInput -= m_PointerInputAction;
                CSHACK_UnBindPointerInput();
                m_PointerInputAction = null;
            }

            if (CSHACK_IsBindedPointerInput())
            {
                CSHACK_PointerInput -= m_PointerInputAction;
                m_PointerInputAction = (UIPointerEvent InEvent) =>
                {
                    if (m_PointerInput != null)
                    {
                        CSHACK_SetPointerInputUIEventResponse(m_PointerInput(InEvent));
                    }
                };
                CSHACK_BindPointerInput();
                CSHACK_PointerInput += m_PointerInputAction;
            }
            else
            {
                m_PointerInputAction = (UIPointerEvent InEvent) =>
                {
                    if (m_PointerInput != null)
                    {
                        CSHACK_SetPointerInputUIEventResponse(m_PointerInput(InEvent));
                    }
                };
                CSHACK_BindPointerInput();
                CSHACK_PointerInput += m_PointerInputAction;
            }
        }

        private void HandleActionInputCPPBindings()
        {
            if (m_ActionInput == null)
            {
                CSHACK_ActionInput -= m_ActionInputAction;
                CSHACK_UnBindActionInput();
                m_ActionInputAction = null;
            }

            if (CSHACK_IsBindedActionInput())
            {
                CSHACK_ActionInput -= m_ActionInputAction;
                m_ActionInputAction = (UIActionEvent InEvent) =>
                {
                    if (m_ActionInput != null)
                    {
                        CSHACK_SetActionInputUIEventResponse(m_ActionInput(InEvent));
                    }
                };
                CSHACK_BindActionInput();
                CSHACK_ActionInput += m_ActionInputAction;
            }
            else
            {
                m_ActionInputAction = (UIActionEvent InEvent) =>
                {
                    if (m_ActionInput != null)
                    {
                        CSHACK_SetActionInputUIEventResponse(m_ActionInput(InEvent));
                    }
                };
                CSHACK_BindActionInput();
                CSHACK_ActionInput += m_ActionInputAction;
            }
        }

        #endregion
    }
}
